Add support for rados_pool_lookup() and rados_pool_reverse_lookup()

nixpanic opened the pull request:
Add missing functions that can be of use for the Ceph-CSI project.

Fixes: #94

Madhu-1 commented on rados/conn.go:
```diff
@@ -280,6 +281,38 @@ func (c *Conn) DeletePool(name string) error {
 	}
 }
 
+// TODO
```
TODO?

nixpanic replied:
Ah, right! Needs docs.

Madhu-1 commented on rados/conn.go:
```diff
@@ -280,6 +281,38 @@ func (c *Conn) DeletePool(name string) error {
+// TODO
+func (c *Conn) GetPoolByName(name string) (int64, error) {
+	if err := c.ensure_connected(); err != nil {
+		fmt.Println("NOT CONNECTED WHOOPS")
```
need to remove  Println

nixpanic replied:
That is just copy/paste from other functions...

Madhu-1 commented on rados/conn.go:
```diff
@@ -280,6 +281,38 @@ func (c *Conn) DeletePool(name string) error {
+func (c *Conn) GetPoolByID(id int64) (string, error) {
+	buf := make([]byte, 4096)
+	if err := c.ensure_connected(); err != nil {
+		fmt.Println("NOT CONNECTED WHOOPS")
```
same here

nixpanic replied:
That is just copy/paste from other functions...

Madhu-1 replied:
Println is not removed yet

Madhu-1 commented on rados/conn.go:
```diff
@@ -280,6 +281,38 @@ func (c *Conn) DeletePool(name string) error {
+	ret := int(C.rados_pool_reverse_lookup(c.cluster, c_id, (*C.char)(unsafe.Pointer(&buf[0])), C.size_t(len(buf))))
+	if ret == -C.ENOENT {
+		return "", RadosError(ret)
+	} else {
```
else is not required as  you are returning in `if` condition

nixpanic replied:
Well, this seems to be the coding practice throughout the file.

Madhu-1 commented on rados/conn.go:
```diff
@@ -280,6 +281,38 @@ func (c *Conn) DeletePool(name string) error {
+	ret := int64(C.rados_pool_lookup(c.cluster, c_name))
+	if ret == -C.ENOENT {
+		return 0, RadosError(ret)
+	} else {
```
else is not required as you are returning in if condition

nixpanic replied:
Well, this seems to be the coding practice throughout the file.

Madhu-1 commented on rados/rados_test.go:
```diff
@@ -296,6 +343,54 @@ func (suite *RadosTestSuite) TestMakeDeletePool() {
 		suite.T().Error("Deleted pool still exists")
 	}
+
+	pool, _ = suite.conn.GetPoolByName(new_name)
```
assert for  error

nixpanic replied:
An error is expected, the pool should not exist anymore.

Madhu-1 commented on rados/rados_test.go:
```diff
@@ -296,6 +343,54 @@ func (suite *RadosTestSuite) TestMakeDeletePool() {
+	assert.NoError(suite.T(), err)
+
+	// verify that it is gone
+	pool, _ = suite.conn.GetPoolByName(new_name)
```
`GetPoolByID`?

Madhu-1 replied:
assert for `err`

nixpanic replied:
Yeah, GetPoolByID() would be better.
An error is expected, the pool should not exist anymore.

nixpanic commented:
On Tue, Oct 15, 2019 at 02:14:55AM -0700, Madhu Rajanna wrote:
> @nixpanic CI failure
> ```
> --- FAIL: TestRadosTestSuite/TestGetPoolByID (1.69s)
>     	rados_test.go:392: Deleted pool still exists
>     --- FAIL: TestRadosTestSuite/TestGetPoolByName (2.03s)
>     	rados_test.go:349: Pool should have been deleted, but was found!
> ```

Oh, yes, I added a check for `err` in the test, obviously the wrong way
around.

nixpanic commented:
@noahdesu care to review this?

nixpanic commented:
On Tue, Oct 22, 2019 at 01:32:53AM -0700, Madhu Rajanna wrote:
> +func (c *Conn) GetPoolByID(id int64) (string, error) {
> +	buf := make([]byte, 4096)
> +	if err := c.ensure_connected(); err != nil {
> +		fmt.Println("NOT CONNECTED WHOOPS")
> 
> Println is not removed yet

No, because other function(s) have that as well. It might be a good
cleanup for some other PR.

phlogistonjohn commented on rados/conn.go:
```diff
@@ -280,6 +281,22 @@ func (c *Conn) DeletePool(name string) error {
+	c_name := C.CString(name)
+	defer C.free(unsafe.Pointer(c_name))
+	ret := int64(C.rados_pool_lookup(c.cluster, c_name))
+	if ret == -C.ENOENT {
```
Is ENOENT the only error condition this api call can return? It seems to me like it would be preferable to handle all error codes as errors at the Go level. Compare the code here to https://github.com/ceph/ceph/blob/master/src/pybind/rados/rados.pyx#L1044 (from the python bindings in ceph/ceph)  it explicitly handles ENOENT but also treates all negative values as the errno. I think we should mimic that behavior.

nixpanic replied:
That is reasonable. I'll replace the check with `ret < 0` and will not handle `ENOENT` differently.

phlogistonjohn commented on rados/conn.go:
```diff
@@ -297,6 +297,22 @@ func (c *Conn) GetPoolByName(name string) (int64, error) {
+	}
+	c_id := C.int64_t(id)
+	ret := int(C.rados_pool_reverse_lookup(c.cluster, c_id, (*C.char)(unsafe.Pointer(&buf[0])), C.size_t(len(buf))))
+	if ret == -C.ENOENT {
```
(My previous comment here vanished) See my comment from the other commit. We need to handle the other error cases here too.
